blog/2024-05-14-migrations-evolution/index.md

```diff
@@ -0,0 +1,119 @@
+In this case, you may need to update the key of the node or edge to include the new property.
+
+### Removing Properties from Nodes and Edges
+
+Conversely, you may also need to remove properties from existing nodes and edges.
+This is equivalent to dropping a column from a table in a relational database.
+When you remove a property from a node or edge, you may need to update existing nodes and edges to remove the value for that property.
+
+### Adding and Removing Indexes
+
+Another common type of schema change is adding and removing indexes.
+Indexes are used to speed up queries by allowing the database to quickly find nodes and edges that match certain criteria.
+When you add an index, you may need to define the properties that the index will be based on and the type of index that will be used.
+When you remove an index, you may need to update existing indexes to remove the properties that the index was based on.
+
+### Topological Changes
+
+Finally, you may need to make topological changes to the schema such as adding or removing relationships between nodes and edges.
+This is equivalent to adding or removing foreign keys in a relational database.
+
+When you change the adjancency of nodes and edges, you may want to clean up the data to ensure that it is consistent with the new schema.
+This may involve updating existing nodes and edges to reflect the new relationships or deleting nodes and edges that are no longer needed.
+
+## How Migrations Work in Nodestream
+
+In nodestream, migrations are defined as a series of yaml files that describe the changes that need to be made to the schema.
+Each migration file contains a list of operations that need to be performed. 
+For example, creating a new node type or adding a property to an existing node type.
+
+When you run `nodestream migrations make` nodestream will create a new migration file in the `migrations` directory. 
+That process works roughly like this:
+
+- Nodestream will look at the current state of the schema by initializing and introspecting all pipelines (A).
+- Build the state of the schema that is represented by the current migrations (B).
+- Diff the two states (A and B) to determine the changes that need to be made to the schema.
+- Generate a new migration file that describes the changes that need to be made to the schema.
+
+When you run `nodestream migrations run` nodestream will apply the migrations in sequence to evolve the schema.
+That process works roughly like this:
+
+- Nodestream reads the migration files into memory and builds a graph of the dependencies between the migrations.
+- Nodestream runs the migrations in topological order, applying the changes to the schema as it goes.
+- Nodestream keeps track of which migrations have been applied so that it can skip them in the future.
+
+Crucially, nodestream does not track all possible schema changes. 
+Topological changes are not tracked, so you will need to handle those manually.
```

Comment:
Its not clear to me what "topological changes" means here.  It might be good to give an example of what would need to be managed outside of the process.